File: mobile/src/main/java/uk/co/appsbystudio/geoshare/friends/adapters/FriendshipStatusAdapter.java
package uk.co.appsbystudio.geoshare.friends.adapters;

import android.content.Context;
import android.content.res.ColorStateList;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.google.firebase.database.FirebaseDatabase;

import java.util.ArrayList;

import de.hdodenhof.circleimageview.CircleImageView;
import uk.co.appsbystudio.geoshare.base.MainActivity;
import uk.co.appsbystudio.geoshare.R;
import uk.co.appsbystudio.geoshare.utils.ProfileUtils;
import uk.co.appsbystudio.geoshare.utils.firebase.listeners.GetUserFromDatabase;

public class FriendshipStatusAdapter extends RecyclerView.Adapter<FriendshipStatusAdapter.ViewHolder>{

    private final Context context;
    private final ArrayList userId;

    public interface Callback{
        void onSendRequest(String friendId);
    }

    private final Callback callback;

    public FriendshipStatusAdapter(Context context, ArrayList userId, Callback callback) {
        this.context = context;
        this.userId = userId;
        this.callback = callback;
    }

    @Override
    public ViewHolder onCreateViewHolder(final ViewGroup viewGroup, final int viewType) {
        final View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.friends_search_item, viewGroup, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        FirebaseDatabase.getInstance().getReference().addListenerForSingleValueEvent(new GetUserFromDatabase(userId.get(position).toString(), holder.friend_name));

        if (!userId.isEmpty()) ProfileUtils.setProfilePicture(userId.get(position).toString(), holder.friends_pictures, context.getCacheDir().toString());

        if (MainActivity.Companion.getFriendsId().containsKey(userId.get(position).toString())) {
            holder.sendRequestButton.setImageDrawable(context.getDrawable(R.drawable.ic_person_white_24dp));
            holder.sendRequestButton.setImageTintList(ColorStateList.valueOf(context.getResources().getColor(R.color.colorPrimary)));
        } else if (MainActivity.Companion.getPendingId().containsKey(userId.get(position).toString())) {
            holder.sendRequestButton.setImageDrawable(context.getDrawable(R.drawable.ic_person_white_24dp));
            holder.sendRequestButton.setImageTintList(ColorStateList.valueOf(context.getResources().getColor(android.R.color.darker_gray)));
        } else {
            holder.sendRequestButton.setImageDrawable(context.getDrawable(R.drawable.ic_send_black_24dp));
            holder.sendRequestButton.setImageTintList(ColorStateList.valueOf(context.getResources().getColor(android.R.color.darker_gray)));
            holder.sendRequestButton.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    callback.onSendRequest(userId.get(holder.getAdapterPosition()).toString());
                    /*holder.sendRequestButton.setImageDrawable(context.getDrawable(R.drawable.ic_person_white_24dp));
                    holder.sendRequestButton.setImageTintList(ColorStateList.valueOf(context.getResources().getColor(android.R.color.darker_gray)));*/
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return userId.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder{

        final TextView friend_name;
        final CircleImageView friends_pictures;
        final ImageButton sendRequestButton;
        final RelativeLayout item;

        ViewHolder(View itemView) {
            super(itemView);
            friend_name = itemView.findViewById(R.id.friend_name);
            friends_pictures = itemView.findViewById(R.id.friend_profile_image);
            sendRequestButton = itemView.findViewById(R.id.sendRequestButton);
            item = itemView.findViewById(R.id.item);
        }
    }
}

File: mobile/src/main/java/uk/co/appsbystudio/geoshare/friends/profile/friends/adapters/FriendshipStatusAdapter.kt
package uk.co.appsbystudio.geoshare.friends.profile.friends.adapters

import android.content.Context
import android.content.res.ColorStateList
import android.support.constraint.ConstraintLayout
import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.TextView
import com.google.firebase.database.FirebaseDatabase
import de.hdodenhof.circleimageview.CircleImageView
import uk.co.appsbystudio.geoshare.R
import uk.co.appsbystudio.geoshare.base.MainActivity
import uk.co.appsbystudio.geoshare.friends.manager.FriendsManager
import uk.co.appsbystudio.geoshare.utils.firebase.FirebaseHelper
import uk.co.appsbystudio.geoshare.utils.firebase.listeners.GetUserFromDatabase
import uk.co.appsbystudio.geoshare.utils.setProfilePicture
import java.util.*

class FriendshipStatusAdapter(private val context: Context?,
                              private val userId: ArrayList<String>,
                              private val callback: Callback?) : RecyclerView.Adapter<FriendshipStatusAdapter.ViewHolder>() {

    interface Callback {
        fun onSendRequest(friendId: String)
    }

    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(viewGroup.context).inflate(R.layout.friends_search_item, viewGroup, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        FirebaseDatabase.getInstance().reference.child("${FirebaseHelper.USERS}/${userId[position]}").addListenerForSingleValueEvent(GetUserFromDatabase(holder.name))

        if (!userId.isEmpty()) holder.profile.setProfilePicture(userId[position], context?.cacheDir.toString())

        when {
            MainActivity.friendsMap.containsKey(userId[position]) -> holder.sendRequestButton.setRequestButton(R.drawable.ic_person_white_24dp, R.color.colorPrimary)
            FriendsManager.pendingUid.containsKey(userId[position]) -> holder.sendRequestButton.setRequestButton(R.drawable.ic_person_white_24dp, android.R.color.darker_gray)
            else -> {
                holder.sendRequestButton.setRequestButton(R.drawable.ic_send_black_24dp, android.R.color.darker_gray)
                holder.sendRequestButton.setOnClickListener {
                    callback?.onSendRequest(userId[holder.adapterPosition])
                }
            }
        }
    }

    private fun ImageButton.setRequestButton(drawable: Int, color: Int) {
        this@setRequestButton.setImageDrawable(context.getDrawable(drawable))
        this@setRequestButton.imageTintList = ColorStateList.valueOf(ResourcesCompat.getColor(context.resources, color, null))
    }

    override fun getItemCount(): Int {
        return userId.size
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        val name: TextView = itemView.findViewById(R.id.text_name_search)
        val profile: CircleImageView = itemView.findViewById(R.id.image_profile_search)
        val sendRequestButton: ImageButton = itemView.findViewById(R.id.button_send_search)
        val item: ConstraintLayout = itemView.findViewById(R.id.item)

    }
}
